### PR TITLE
refactor: upgrade kube client version to v0.19.6. Fixes #4425, #4791

### LINKED ISSUE #4425: Update k8s.io/api, k8s.io/apimachinery and k8s.io/client-go
# Summary

Update go dependencies to stay up to date. 

# Use Cases

I have had issues trying to updated the argo version for Kubeflow pipelines to argo 2.11(currently are on 2.7) and updating the underlying dependencies for argo would make the update simpler. I understand this might not be a good enough reason to update the dependencies for argo it self though. 

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### LINKED ISSUE #4791: Broken import when using Argo go client
## Summary 

When building a service that invokes Argo using the go client, I frequently run into the [gnostic openapiv2 error](https://github.com/kubernetes/client-go/issues/741). [Newer kube dependencies](https://github.com/kubernetes/kube-openapi/pull/189) seem to have fixed this, but [existing versions](https://github.com/kubernetes/client-go/issues/741#issuecomment-606883355) will not be fixed 😞. 

I'd propose bumping the K8s dependencies to a newer version. I'm happy to own the issue and make the changes.
 
## Diagnostics

Here are my relevant go dependencies in go.mod:
```
require (
	github.com/argoproj/argo v0.0.0-20200724211756-20d2ace3d534
	k8s.io/api v0.17.8
	k8s.io/apimachinery v0.17.8
	k8s.io/client-go v0.17.8
	k8s.io/kubectl v0.17.5
)
```
Note: I use the scheme package in kubectl to interact with secrets.

Here's the error I get when running a build:
```
cannot load github.com/googleapis/gnostic/OpenAPIv2: module github.com/googleapis/gnostic@latest found (v0.5.3), but does not contain package github.com/googleapis/gnostic/OpenAPIv2
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by JPZ13**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Running e2e tests here because my machine times out before completion

This PR:
- Upgrades dependencies on kube (api, apimachinery, client-go, etc) to v0.20.1
- Updates kube invocations with updated function signatures (lots of context and metav1 options)
- Closes #4791 and #4425

### CODE REVIEW

**Comment by JPZ13** [comment]
@alexec The log output doesn't say "Config loaded from file" anymore. Would changing it be acceptable? Also, running `make codegen` on my system isn't fixing the CI codegen. Do you have suggestions on how to fix?

cc @NikeNano

**Comment by NikeNano** [comment]
> ? Also, running `make codegen` on my system isn't fixing the CI codegen. Do you have suggestions on how to fix?

Are you using the same version of swagger as argo currently uses? Seems like the definition don't follow version 2.

**Comment by JPZ13** [comment]
> > Also, running `make codegen` on my system isn't fixing the CI codegen. Do you have suggestions on how to fix?
> 
> Are you using the same version of swagger as argo currently uses @JPZ13 ? Seems like the definition don't follow version 2.

@NikeNano I installed it using the `make` command in the argo root directory, but nothing changes. Here's my terminal output:
```
$ make $GOPATH/bin/swagger
make: `/Users/jpz13/go/bin/swagger' is up to date.

$ which swagger
/Users/jpz13/go/bin/swagger

$ swagger version
dev

$ make codegen
env ARGO_SECURE=false ARGO_INSECURE_SKIP_VERIFY=false ARGO_SERVER= ARGO_INSTANCEID= go run ./hack docgen
generating docs/fields.md
go run ./hack/cli
generating docs/cli
# `go generate ./...` takes around 10s, so we only run on specific packages.
go generate ./persist/sqldb ./pkg/apiclient/workflow ./server/auth ./server/auth/sso ./workflow/executor
Generating mock for: WorkflowArchive in file: mocks/WorkflowArchive.go
Generating mock for: WorkflowServiceClient in file: mocks/WorkflowServiceClient.go
Generating mock for: Gatekeeper in file: mocks/Gatekeeper.go
Generating mock for: Interface in file: mocks/Interface.go
Generating mock for: ContainerRuntimeExecutor in file: mocks/ContainerRuntimeExecutor.go
rm -Rf vendor

$ git status
On branch upgrade-kube
nothing to commit, working tree clean
```

**Comment by NikeNano** [comment]
hmm looks strange with:

```
$ swagger version
dev
```
maybe if you try to install version 2.X. 

Since the error for `CI / Codegen` mentions:
```
The swagger spec at "api/openapi-spec/swagger.json" is invalid against swagger specification 2.0. 
```

but I don't know these things ..... so guess we have to wait for @alexec to get feedback on this.

**Comment by alexec** [comment]
* We will want/need to have the same Kubernetes client version as Argo Events. Can you please look at https://github.com/argoproj/argo-events/issues/999#issuecomment-749199296 ?
* You can force recreation of files as follows `rm api/openapi-spec/swagger.json && make api/openapi-spec/swagger.json`. I'm not sure what the error is.
* I think it would be absolutely fine to use `context.TODO()` to make things simpler, but I would not go back and change anything now

**Comment by JPZ13** [comment]
@alexec Sounds good. I switched the dependencies to match Argo events. To fix the test-cli error, I think we have to upgrade the k8s dependencies [here](https://github.com/argoproj/pkg/blob/master/go.mod#L17). That will fix klog. Should I skip those in the meantime and make the update to the argoproj/pkg repo?

I still can't get codegen to fix despite doing what you suggested. Feel free to pull this branch, make a commit running `make codegen`, and push that commit.

Also, happy new year. I'm game to pause on this and resume on the 4th if you're taking some time to relax.

**Comment by alexec** [comment]
Looks good. I'll ask Derek to review as he did the work for Argo Events. You need to make these changes:

## test-e2e

TestGLogLevels is failiing.

I would say change `k8s.io/klog` to `k8s.io/klog/v2` in `cli.go`, but that file is part of `argoproj/pkg` and cannot be changed.

Instead,  change `cmd/argo/commands/root.go` as fal/ows:

```golang
	command.PersistentPreRun = func(cmd *cobra.Command, args []string) {
		if verbose {
			logLevel = "debug"
			glogLevel = 6
		}
		cli.SetLogLevel(logLevel)
		cli.SetGLogLevel(glogLevel)
                // INSERT THESE LINES:
		klog.InitFlags(nil).   
		_ = flag.Set("logtostderr", "true")
		_ = flag.Set("v", strconv.Itoa(glogLevel))
                // END 
		log.WithField("version", argo.GetVersion()).Debug("CLI version")
	}
```

## codegen

This is failing Swagger validation. I don't think the `swagger` version being `dev` is the issue - this is my version locally. 

My first hypothesis - the underlying problem is because `controller-tools` is pinned to v0.2.9 in `go.mod`. Delete lines 93..96, run `go mod tidy` then `touch touch pkg/apis/workflow/v1alpha1/workflow_types.go` (this will force many of the codegen targets to run), then`make codegen `, finally commit and push the changes.

If that does not work, I'd like to know how the swagger.json has change. Insert `git diff` before `swagger validate` in the `Makefile`;

```
api/openapi-spec/swagger.json: $(GOPATH)/bin/swagger dist/kubeified.swagger.json
	swagger flatten --with-flatten minimal --with-flatten remove-unused dist/kubeified.swagger.json -o api/openapi-spec/swagger.json
        # INSERT THIS LINE:
        git diff
        # END 
	swagger validate api/openapi-spec/swagger.json
	go test ./api/openapi-spec
```

**Review by alexec — CHANGES_REQUESTED** [review]
See comment.

**Comment by JPZ13** [comment]
Thanks for the review @alexec. I deleted the lines in the `go.mod` file, ran `go mod tidy`, and did the other steps with no luck. Here's the output for `git diff` after the `swagger flatten` command:
```
diff --git a/api/openapi-spec/swagger.json b/api/openapi-spec/swagger.json
index 99b43c0e..73cae18d 100644
--- a/api/openapi-spec/swagger.json
+++ b/api/openapi-spec/swagger.json
@@ -4154,10 +4154,6 @@
           "description": "Arguments extracted from the event and then set as arguments to the workflow created.",
           "$ref": "#/definitions/io.argoproj.workflow.v1alpha1.Arguments"
         },
-        "metadata": {
-          "description": "Metadata optional means to customize select fields of the workflow metadata",
-          "$ref": "#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta"
-        },
         "workflowTemplateRef": {
           "description": "WorkflowTemplateRef the workflow template to submit",
           "$ref": "#/definitions/io.argoproj.workflow.v1alpha1.WorkflowTemplateRef"
```

**Comment by alexec** [comment]
You have a panic in your tests:
```
[36margo-server | [m./dist/argo flag redefined: log_dir
[36margo-server | [mpanic: ./dist/argo flag redefined: log_dir
[36margo-server | [mgoroutine 1 [running]:
[36margo-server | [mflag.(*FlagSet).Var(0xc0000a0180, 0x278a660, 0x3982150, 0x22a6bf3, 0x7, 0x23049ec, 0x2f)
[36margo-server | [m	/opt/hostedtoolcache/go/1.13.12/x64/src/flag/flag.go:848 +0x4ae
[36margo-server | [mflag.(*FlagSet).StringVar(...)
[36margo-server | [m	/opt/hostedtoolcache/go/1.13.12/x64/src/flag/flag.go:751
[36margo-server | [mk8s.io/klog.InitFlags(0x0)
[36margo-server | [m	/home/runner/go/pkg/mod/k8s.io/klog@v1.0.0/klog.go:420 +0x7f
[36margo-server | [mgithub.com/argoproj/pkg/cli.SetGLogLevel(0x0)
[36margo-server | [m	/home/runner/go/pkg/mod/github.com/argoproj/pkg@v0.3.0/cli/cli.go:22 +0x2e
[36margo-server | [mgithub.com/argoproj/argo/cmd/argo/commands.NewCommand.func2(0xc0005badc0, 0xc0003c3180, 0x0, 0x8)
[36margo-server | [m	/home/runner/work/argo/argo/cmd/argo/commands/root.go:128 +0xa7
[36margo-server | [mgithub.com/spf13/cobra.(*Command).execute(0xc0005badc0, 0xc0003c3100, 0x8, 0x8, 0xc0005badc0, 0xc0003c3100)
[36margo-server | [m	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:826 +0x21d
[36margo-server | [mgithub.com/spf13/cobra.(*Command).ExecuteC(0xc000647080, 0xc000000180, 0xc000745f50, 0x4073cf)
[36margo-server | [m	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
[36margo-server | [mgithub.com/spf13/cobra.(*Command).Execute(...)
[36margo-server | [m	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
[36margo-server | [mmain.main()
[36margo-server | [m	/home/runner/work/argo/argo/cmd/argo/main.go:13 +0x2b
[36margo-server | [mTerminating argo-server
```

**Comment by alexec** [comment]
can you try this?

If that does not fix this issue (or reveal the underlying cause), I'd like to know how the swagger.json has change. Insert git diff before swagger validate in the Makefile;

```
api/openapi-spec/swagger.json: $(GOPATH)/bin/swagger dist/kubeified.swagger.json
	swagger flatten --with-flatten minimal --with-flatten remove-unused dist/kubeified.swagger.json -o api/openapi-spec/swagger.json
        # INSERT THIS LINE:
        git diff
        # END 
	swagger validate api/openapi-spec/swagger.json
	go test ./api/openapi-spec
```

**Comment by alexec** [comment]
I look like you need to change your Makefile line 134 to:

-I ${GOPATH}/pkg/mod/github.com/grpc-ecosystem/grpc-gateway@v1.16.0/third_party/googleapis \

**Comment by JPZ13** [comment]
I'm pretty stuck @alexec. Here's what I've tried so far:
- removing lines 93 through 96 in the `go.mod` file
- re-downloading all of the `make codegen` dependencies like protoc-gen-gogo
- re-running `make codegen`

I've also placed a `git diff` where you requested. It seems all of the other tests are passing, including the previously failing cli test. Let me know if you think of anything else I should try.

**Comment by alexec** [comment]
Tip: you can force always remake targets (@simster7 @sarabala1979 @whynowy).

```
git clean -fxd  ;# delete all files not committed to make cGit 
make -B codegen ;# make all targets (including installing deps) - this will take a long time
````

**Comment by alexec** [comment]
I'm investigating... I think this is because 

1. `protoc` is piped to `grep` by does not output any lines so `grep` fails - this `grep` can be removed
2. CRDs have changed spec and we need some new code

**Comment by alexec** [comment]
I'm wrong. Swagger v3.0 introduces 'default'. We need to use the correct `swagger`

**Comment by alexec** [comment]
Not possible: they are refusing to upgrade to support v3.0: https://github.com/go-swagger/go-swagger/issues/1122

**Comment by JPZ13** [comment]
Thanks for the tip, and thanks for diving into this @alexec

**Comment by alexec** [comment]
I think we can pin `openapi-gen`. It's not used in production, only for generating `pkg/apis/workflow/v1alpha1/openapi_generated.go`...

**Comment by alexec** [comment]
I think you can apply these diffs:

```
diff --git a/Makefile b/Makefile
index 9c9ca82d..52b71ee2 100644
--- a/Makefile
+++ b/Makefile
@@ -135,7 +135,7 @@ define protoc
       --gogofast_out=plugins=grpc:${GOPATH}/src \
       --grpc-gateway_out=logtostderr=true:${GOPATH}/src \
       --swagger_out=logtostderr=true,fqn_for_swagger_name=true:. \
-      $(1) 2>&1 | grep -v 'warning: Import .* is unused'
+      $(1)
 endef
 # docker_build,image_name,binary_name,marker_file_name
 define docker_build
@@ -533,7 +533,6 @@ dist/kubeified.swagger.json: dist/swaggifed.swagger.json dist/kubernetes.swagger
 
 api/openapi-spec/swagger.json: $(GOPATH)/bin/swagger dist/kubeified.swagger.json
 	swagger flatten --with-flatten minimal --with-flatten remove-unused dist/kubeified.swagger.json -o api/openapi-spec/swagger.json
-	git diff
 	swagger validate api/openapi-spec/swagger.json
 	go test ./api/openapi-spec
 
diff --git a/hack/crds.go b/hack/crds.go
index b4d34814..27ff7df6 100644
--- a/hack/crds.go
+++ b/hack/crds.go
@@ -20,7 +20,8 @@ func cleanCRD(filename string) {
 	metadata := crd["metadata"].(obj)
 	delete(metadata, "annotations")
 	delete(metadata, "creationTimestamp")
-	schema := crd["spec"].(obj)["validation"].(obj)["openAPIV3Schema"].(obj)
+	versions := crd["spec"].(obj)["versions"].([]interface{})
+	schema := versions[0].(obj)["schema"].(obj)["openAPIV3Schema"].(obj)
 	name := crd["metadata"].(obj)["name"].(string)
 	switch name {
 	case "cronworkflows.argoproj.io":
diff --git a/hack/swagger/kubeifyswagger.go b/hack/swagger/kubeifyswagger.go
index e50c4399..0520a955 100644
--- a/hack/swagger/kubeifyswagger.go
+++ b/hack/swagger/kubeifyswagger.go
@@ -35,12 +35,14 @@ func kubeifySwagger(in, out string) {
 
 	//loop again to handle any new bad definitions
 	for _, d := range definitions {
-		if d.(obj)["properties"] != nil {
-			props := d.(obj)["properties"].(obj)
+		props, ok := d.(obj)["properties"].(obj)
+		if ok {
 			for _, content := range props {
-				if content.(obj)["format"] == "int32" || content.(obj)["format"] == "int64" {
-					delete(content.(obj), "format")
+				prop := content.(obj)
+				if prop["format"] == "int32" || prop["format"] == "int64" {
+					delete(prop, "format")
 				}
+				delete(prop, "default")
 			}
 		}
 	}

```

Then run 

```
make codegen -B
````

**Comment by JPZ13** [comment]
Awesome. I'm on a call right now but will add them in about 45 minutes. Thanks @alexec

**Comment by alexec** [comment]
I think you need to sync with master?

**Comment by JPZ13** [comment]
> I think you need to sync with master?

Just did - it took a quick second to rebase

**Comment by JPZ13** [comment]
apimachinery's ToUnstructured function can't handle the inline change from #4828. I'll fill out an issue on apimachinery tomorrow

**Comment by alexec** [comment]
We may need to remove `writeBackToInformer`. I think you still need to edit the `Makefile` to delete this:

```
2>&1 | grep -v 'warning: Import .* is unused'
```

**Comment by JPZ13** [comment]
This is the diff that fixed codegen. It was failing because the schema of the crds changed. My bad for not really understanding the `/hack` directory
```
diff --git a/hack/crds.go b/hack/crds.go
index 27ff7df6..7ee684d5 100644
--- a/hack/crds.go
+++ b/hack/crds.go
@@ -20,8 +20,9 @@ func cleanCRD(filename string) {
        metadata := crd["metadata"].(obj)
        delete(metadata, "annotations")
        delete(metadata, "creationTimestamp")
-       versions := crd["spec"].(obj)["versions"].([]interface{})
-       schema := versions[0].(obj)["schema"].(obj)["openAPIV3Schema"].(obj)
+       spec := crd["spec"].(obj)
+       validation := spec["validation"].(obj)
+       schema := validation["openAPIV3Schema"].(obj)
        name := crd["metadata"].(obj)["name"].(string)
        switch name {
        case "cronworkflows.argoproj.io":
```

I'll remove `writeBackToInformer` next

File: workflow/util/unstructured.go
Comment (alexec) [inline workflow/util/unstructured.go:5]: it would be easier to revert https://github.com/argoproj/argo/pull/4828 yes?

Comment (JPZ13) [reply workflow/util/unstructured.go:5]: Probably - I just didn't know if you guys wanted to do that

Comment (alexec) [reply workflow/util/unstructured.go:5]: this is a trade-off, and I think reverting it would be safer, even if it is a shame

Comment (JPZ13) [reply workflow/util/unstructured.go:5]: Makes sense. I think it's a shame as well

Comment (JPZ13) [reply workflow/util/unstructured.go:5]: Tagging @taharah as a heads up. The inlining is preventing an upgrade to apimachinery. Going to revert your commit to safely accommodate apimachinery v0.19.6 which kicks a panic when `ParallelSteps` are inlined 😞

Comment (taharah) [reply workflow/util/unstructured.go]: Thanks for tagging me. I'll checkout your changes and see if I can get it working again.

File: cmd/argo/commands/root.go
Comment (alexec) [inline cmd/argo/commands/root.go:6]: minor: needs goimport

File: workflow/cron/controller.go
Comment (alexec) [inline workflow/cron/controller.go:99]: TODO

File: workflow/templateresolution/context.go
Comment (alexec) [inline workflow/templateresolution/context.go:68]: TODO

File: workflow/templateresolution/context.go
Comment (alexec) [inline workflow/templateresolution/context.go:34]: TODO - i.e. `context.TODO()`

**Review by alexec — COMMENTED** [review]
Some minor comments. @sarabala1979 can I please ask you to review this? It's mostly boiler-plate insertion of context in the code. I just want to make sure nothing accidentally appeared in it.

File: workflow/util/unstructured.go
Comment (JPZ13) [reply workflow/util/unstructured.go]: Thanks @taharah!

File: cmd/argo/commands/root.go
Comment (JPZ13) [reply cmd/argo/commands/root.go:6]: These packages are used on line 147

File: workflow/cron/controller.go
Comment (JPZ13) [reply workflow/cron/controller.go]: Updated with context.TODO()

File: workflow/templateresolution/context.go
Comment (JPZ13) [reply workflow/templateresolution/context.go]: Updated with context.TODO()

File: workflow/templateresolution/context.go
Comment (JPZ13) [reply workflow/templateresolution/context.go]: Updated with context.TODO()

**Comment by alexec** [comment]
@sarabala1979 @simster7 could I get another pair of eyes on this PR please?

File: cmd/argo/commands/root.go
Comment (alexec) [reply cmd/argo/commands/root.go:6]: more that the imports are not sorted and grouped - but this is minor

File: cmd/argo/commands/root.go
Comment (alexec) [inline cmd/argo/commands/root.go:144]: this needs to be fixed in three places in total:

* here
* cmd/argoexec/commands/root.go
* cmd/workflow-controller/main.go

**Review by alexec — CHANGES_REQUESTED** [review]
I've re-reviewed - just one final change please!

File: cmd/argo/commands/root.go
Comment (JPZ13) [reply cmd/argo/commands/root.go]: Ah, makes sense. I deleted the line breaks that were there previously and re-ran goimport for this new commit

File: cmd/argo/commands/root.go
Comment (JPZ13) [reply cmd/argo/commands/root.go]: Sounds good - I moved the function to the `util/cmd` package and then invoked it in each of the files you listed

**Comment by JPZ13** [comment]
> I've re-reviewed - just one final change please!

All good!

File: util/cmd/cmd.go
Comment (alexec) [inline util/cmd/cmd.go:14]: wrong klog

File: util/cmd/cmd.go
Comment (alexec) [inline util/cmd/cmd.go:101]: minor - could go into own file

File: util/cmd/cmd.go
Comment (alexec) [reply util/cmd/cmd.go:14]: this is why tests are failing again I think

Comment (JPZ13) [reply util/cmd/cmd.go]: Yeah my bad on that - pushing the commit now

File: util/cmd/cmd.go
Comment (JPZ13) [reply util/cmd/cmd.go:100]: Sounds good - i'll scoot it into a separate file real quick

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
This is ready to merge, I think we now need to get v1.2 of Argo Events released so that we can import it on the  `ui-v3` branch. I'll speak to @whynowy about time-frames tomorrow.

**Comment by simster7** [comment]
> @sarabala1979 @simster7 could I get another pair of eyes on this PR please?

Will take a look today just for a sanity check :)

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:437]: Is this intended? I'm not certain, but this could be a big change?

File: workflow/controller/controller.go
Comment (simster7) [inline workflow/controller/controller.go:232]: Should this be an issue instead? I've found that `// TODOs` in the codebase just stay there and nothing ever gets done. Why not open a new issue and assign to yourself? Then we can remove this comment

File: workflow/controller/controller.go
Comment (simster7) [inline workflow/controller/controller.go:420]: Is this formating from a linter or did you do this yourself? Just wondering

File: workflow/controller/dag.go
Comment (simster7) [inline workflow/controller/dag.go:348]: Hmmm... in this function we add a `ctx` to a bunch of signatures just for this method. Makes me wonder if we should just save the context in `woc.ctx`? That way we don't have to pass it in a huge stack of functions just for one or two uses. @alexec What do you think?

File: workflow/controller/steps.go
Comment (simster7) [inline workflow/controller/steps.go:271]: Same here, for example

File: workflow/cron/controller.go
Comment (simster7) [inline workflow/cron/controller.go:233]: Why are we using background here? I understand that the way `syncAll` gets called makes it so you can't pass parameters, but it would still be possible to pass it in using a higher-level function definition.

Maybe we should also add a `cc.ctx` or at least use a high-level def

**Review by simster7 — CHANGES_REQUESTED** [review]
Thanks for doing all this work! My main question is whether we should consider saving the context on the structs instead of passing it down function signatures.

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:437]: this recent change (2-3 day old) change needed to be reverted because it does not work with the new code

File: workflow/controller/dag.go
Comment (alexec) [reply workflow/controller/dag.go:348]: AFAIK you should rarely have `Context` as a field, it is intended to be passed around as the first parameter to functions. But I get that is is a bit onerous.

Comment (JPZ13) [reply workflow/controller/dag.go:348]: I'm down with whatever y'all want to go with, but my suggestion would be to keep it as the first argument in functions because it's more idiomatic: https://blog.golang.org/context

File: workflow/controller/controller.go
Comment (JPZ13) [reply workflow/controller/controller.go:420]: I don't remember. I think I did it manually to make it easier to read given that the arguments list is pretty long

File: workflow/cron/controller.go
Comment (JPZ13) [reply workflow/cron/controller.go:233]: This one might just be a typo where I didn't see that `ctx` is declared in the function signature. I'll remove line 233 here

File: workflow/controller/controller.go
Comment (JPZ13) [reply workflow/controller/controller.go:232]: Good catch. I'll remove the `TODO` and do a `grep` for any others that I put in and file issues instead

**Comment by JPZ13** [comment]
> Thanks for doing all this work! My main question is whether we should consider saving the context on the structs instead of passing it down function signatures.

Thanks @simster7. I appreciate it. I'll make the changes you mentioned in the inline issues tomorrow

**Comment by JPZ13** [comment]
The most recent commit addresses your line-item changes @simster7. Let me know if it all looks good

**Comment by alexec** [comment]
@simster7 do you want to review again? I'm happy to merge.

**Review by simster7 — APPROVED** [review]

**Comment by simster7** [comment]
@alexec good to merge

**Comment by alexec** [comment]
Next step after this is to change `go.mod`  to contain `github.com/argoproj/argo/v3'. Let's push things forward!

**Pull request merged**
